docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1056 @@
+The fourth method expects only a managed reference (`ref T source`). We don't need to pin the buffer (GC is tracking managed references and updates them if memory gets moved), but it still requires us to properly handle managed pointer arithmetic:
+
+```cs
+int ManagedReferencesSum(int[] buffer)
+{
+    ref int current = ref MemoryMarshal.GetArrayDataReference(buffer);
+    ref int end = ref Unsafe.Add(ref current, buffer.Length);
+    ref int oneVectorAwayFromEnd = ref Unsafe.Subtract(ref end, Vector128<int>.Count);
+
+    Vector128<int> sum = Vector128<int>.Zero;
+
+    while (!Unsafe.IsAddressGreaterThan(ref current, ref oneVectorAwayFromEnd))
+    {
+        sum += Vector128.LoadUnsafe(ref current);
+
+        current = ref Unsafe.Add(ref current, Vector128<int>.Count);
+    }
+
+    int result = Vector128.Sum(sum);
+
+    while (Unsafe.IsAddressLessThan(ref current, ref end))
+    {
+        result += current;
+
+        current = ref Unsafe.Add(ref current, 1);
+    }
+
+    return result;
+}
+```
+
+**Note:** `Unsafe` does not expose a method called "IsGreaterOrEqualThan", so we are using a negation of `Unsafe.IsAddressGreaterThan` to achieve desired effect.
+
+**Pointer arithmetic can always go wrong, even if you are an experienced engineer and get a very detailed code review from .NET architects**. In [#73768](https://github.com/dotnet/runtime/pull/73768) a GC hole was introduced. The code looked simple:
+
+```cs
+ref TValue currentSearchSpace = ref Unsafe.Add(ref searchSpace, length - Vector128<TValue>.Count);
+
+do
+{
+    equals = Vector128.Equals(values, Vector128.LoadUnsafe(ref currentSearchSpace));
+    if (equals == Vector128<TValue>.Zero)
+    {
+        currentSearchSpace = ref Unsafe.Subtract(ref currentSearchSpace, Vector128<TValue>.Count);
+        continue;
+    }
+
+    return ...;
+}
+while (!Unsafe.IsAddressLessThan(ref currentSearchSpace, ref searchSpace));
+```
+
+It was part of `LastIndexOf` implementation, where we were iterating from the end to the beginning of the buffer. In the last iteration of the loop, `currentSearchSpace` could become a pointer to unknown memory that lied before the beginning of the buffer:
+
+```cs
+currentSearchSpace = ref Unsafe.Subtract(ref currentSearchSpace, Vector128<TValue>.Count);
+```
+
+And it was fine until GC kicked right after that, moved objects in memory, updated all valid managed references and resumed the execution, which run following condition:
+
+```cs
+while (!Unsafe.IsAddressLessThan(ref currentSearchSpace, ref searchSpace));
+```
+
+Which could return true because `currentSearchSpace` was invalid and not updated. If you are interested in more details, you can check the [issue](https://github.com/dotnet/runtime/issues/75792#issuecomment-1249973858) and the [fix](https://github.com/dotnet/runtime/pull/75857).
+
+That is why **we recommend using the overload that takes a managed reference and an element offset. It does not require pinning or doing any pointer arithmetic!**
```

Comment:
But still requires care; you can still pass in an incorrect offset and get a GC hole.